activerecord/lib/active_record/insert_all.rb

```diff
@@ -69,14 +76,17 @@ def readonly_columns
       end
 
       def unique_by_columns
-        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+        Array(unique_by_index&.columns)
       end
 
-      def primary_keys
-        Array.wrap(model.primary_key)
+      def find_index_for(unique_by_index)
+        if index = connection.indexes(model.table_name).index_by(&:name)[unique_by_index.to_s]
```

Comment:
@kamipo is this okay to do? Or will it execute a query for the schema?

Reply:
Another thing we could do is matching on the columns ala: `find { |index| index.columns.sort == unique_by_index }` to get `unique_by_index: :isbn` and `unique_by_index: %i( author_id name )`.

Reply:
I think it is a good timing to make the index cache on `schema_cache`.
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/schema_cache.rb

Reply:
Great, I gotta try that!